File: carpool/backend/routes/bookings.js
const express = require('express');
const router  = express.Router();
const db      = require('../db/connection');
const { authenticate, requirePassenger } = require('../middleware/auth');

// POST /bookings — passenger books a trip
router.post('/', authenticate, requirePassenger, async (req, res) => {
  const { trip_id, seats_booked } = req.body;

  if (!trip_id || !seats_booked) {
    return res.status(400).json({ error: 'trip_id and seats_booked are required.' });
  }
  if (Number(seats_booked) <= 0) {
    return res.status(400).json({ error: 'seats_booked must be at least 1.' });
  }

  const conn = await db.getConnection();
  try {
    await conn.beginTransaction();

    // Lock trip row for update (prevent race conditions)
    const [trips] = await conn.query(
      'SELECT * FROM trips WHERE trip_id = ? AND status = "active" FOR UPDATE',
      [trip_id]
    );
    if (trips.length === 0) {
      await conn.rollback();
      return res.status(404).json({ error: 'Trip not found or not active.' });
    }
    const trip = trips[0];

    // Prevent passenger from booking their own trip
    if (trip.driver_id === req.user.user_id) {
      await conn.rollback();
      return res.status(400).json({ error: 'Drivers cannot book their own trips.' });
    }

    // Check for duplicate booking
    const [existing] = await conn.query(
      'SELECT booking_id FROM bookings WHERE trip_id = ? AND passenger_id = ? AND status != "cancelled"',
      [trip_id, req.user.user_id]
    );
    if (existing.length > 0) {
      await conn.rollback();
      return res.status(409).json({ error: 'You already have an active booking for this trip.' });
    }

    // Overbooking check
    if (trip.available_seats < Number(seats_booked)) {
      await conn.rollback();
      return res.status(400).json({
        error: `Not enough seats. Only ${trip.available_seats} seat(s) available.`
      });
    }

    // Create booking
    const [bookingResult] = await conn.query(
      'INSERT INTO bookings (trip_id, passenger_id, seats_booked, status) VALUES (?, ?, ?, "confirmed")',
      [trip_id, req.user.user_id, seats_booked]
    );

    // Deduct seats immediately
    await conn.query(
      'UPDATE trips SET available_seats = available_seats - ? WHERE trip_id = ?',
      [seats_booked, trip_id]
    );

    // Auto-create a pending payment record
    const amount = trip.price * Number(seats_booked);
    await conn.query(
      'INSERT INTO payments (booking_id, amount, payment_status) VALUES (?, ?, "pending")',
      [bookingResult.insertId, amount]
    );

    await conn.commit();

    res.status(201).json({
      message: 'Booking confirmed.',
      booking: {
        booking_id: bookingResult.insertId,
        trip_id,
        passenger_id: req.user.user_id,
        seats_booked,
        status: 'confirmed',
        amount_due: amount
      }
    });
  } catch (err) {
    await conn.rollback();
    console.error(err);
    res.status(500).json({ error: 'Server error.' });
  } finally {
    conn.release();
  }
});

// GET /bookings — get current user's bookings
router.get('/', authenticate, async (req, res) => {
  try {
    const [bookings] = await db.query(`
      SELECT 
        b.*,
        t.origin, t.destination, t.departure_time, t.price,
        u.name AS driver_name,
        p.amount, p.payment_status
      FROM bookings b
      JOIN trips t ON t.trip_id = b.trip_id
      JOIN users u ON u.user_id = t.driver_id
      LEFT JOIN payments p ON p.booking_id = b.booking_id
      WHERE b.passenger_id = ?
      ORDER BY b.created_at DESC
    `, [req.user.user_id]);
    res.json({ bookings });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Server error.' });
  }
});

// DELETE /bookings/:id — passenger cancels a booking
router.delete('/:id', authenticate, requirePassenger, async (req, res) => {
  const conn = await db.getConnection();
  try {
    await conn.beginTransaction();

    // Verify booking belongs to user
    const [rows] = await conn.query(
      'SELECT * FROM bookings WHERE booking_id = ? AND passenger_id = ?',
      [req.params.id, req.user.user_id]
    );
    if (rows.length === 0) {
      await conn.rollback();
      return res.status(404).json({ error: 'Booking not found.' });
    }
    const booking = rows[0];
    if (booking.status === 'cancelled') {
      await conn.rollback();
      return res.status(400).json({ error: 'Booking is already cancelled.' });
    }

    // Cancel booking
    await conn.query(
      'UPDATE bookings SET status = "cancelled" WHERE booking_id = ?',
      [booking.booking_id]
    );

    // Restore seats
    await conn.query(
      'UPDATE trips SET available_seats = available_seats + ? WHERE trip_id = ?',
      [booking.seats_booked, booking.trip_id]
    );

    await conn.commit();
    res.json({ message: 'Booking cancelled and seats restored.' });
  } catch (err) {
    await conn.rollback();
    console.error(err);
    res.status(500).json({ error: 'Server error.' });
  } finally {
    conn.release();
  }
});

module.exports = router;

File: backend/app/routes/bookings.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import List
from ..models.base import get_db
from ..models.schemas import Booking, Trip, User

router = APIRouter()


class BookingCreate(BaseModel):
    trip_id: int
    passenger_id: int
    seats_booked: int


@router.post("/", status_code=status.HTTP_201_CREATED)
def create_booking(booking_data: BookingCreate, db: Session = Depends(get_db)):
    trip = db.query(Trip).filter(Trip.id == booking_data.trip_id).first()
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")

    if trip.status != "active":
        raise HTTPException(status_code=400, detail="This trip is not available for booking")

    if trip.driver_id == booking_data.passenger_id:
        raise HTTPException(status_code=400, detail="Drivers cannot book their own trips")

    if booking_data.seats_booked <= 0:
        raise HTTPException(status_code=400, detail="Seats booked must be greater than 0")

    if trip.available_seats < booking_data.seats_booked:
        raise HTTPException(status_code=400, detail="Not enough seats available")

    passenger = db.query(User).filter(User.id == booking_data.passenger_id).first()
    if not passenger:
        raise HTTPException(status_code=404, detail="Passenger not found")

    total_price = trip.price_per_seat * booking_data.seats_booked

    if passenger.balance < total_price:
        raise HTTPException(
            status_code=400,
            detail=f"Insufficient balance. Required: {total_price}, Current: {passenger.balance}",
        )

    # Deduct balance immediately upon booking (escrow-like)
    passenger.balance -= total_price

    # DEDUCT SEATS IMMEDIATELY to prevent overbooking
    trip.available_seats -= booking_data.seats_booked

    new_booking = Booking(
        trip_id=booking_data.trip_id,
        passenger_id=booking_data.passenger_id,
        seats_booked=booking_data.seats_booked,
        total_price=total_price,
        status="paid",  # Starts as paid because balance is deducted
    )
    db.add(new_booking)
    db.commit()
    db.refresh(new_booking)
    return new_booking


@router.get("/user/{user_id}")
def get_user_bookings(user_id: int, db: Session = Depends(get_db)):
    bookings = db.query(Booking).filter(Booking.passenger_id == user_id).all()
    results = []
    for b in bookings:
        trip = db.query(Trip).filter(Trip.id == b.trip_id).first()
        results.append(
            {
                "id": b.id,
                "trip": {
                    "id": trip.id,
                    "origin": trip.origin,
                    "destination": trip.destination,
                    "departure_time": trip.departure_time,
                    "price_per_seat": trip.price_per_seat,
                },
                "seats_booked": b.seats_booked,
                "total_price": b.total_price,
                "status": b.status,
                "created_at": b.created_at,
            }
        )
    return results


@router.get("/driver/{driver_id}")
def get_driver_bookings(driver_id: int, db: Session = Depends(get_db)):
    # Join with Trip to filter by driver_id
    bookings = db.query(Booking).join(Trip).filter(Trip.driver_id == driver_id).all()
    results = []
    for b in bookings:
        passenger = db.query(User).filter(User.id == b.passenger_id).first()
        trip = db.query(Trip).filter(Trip.id == b.trip_id).first()
        results.append(
            {
                "id": b.id,
                "trip": {
                    "id": trip.id,
                    "origin": trip.origin,
                    "destination": trip.destination,
                    "departure_time": trip.departure_time,
                },
                "passenger": {"id": passenger.id, "full_name": passenger.full_name},
                "seats_booked": b.seats_booked,
                "total_price": b.total_price,
                "status": b.status,
            }
        )
    return results


@router.get("/{booking_id}")
def get_booking_detail(booking_id: int, db: Session = Depends(get_db)):
    booking = db.query(Booking).filter(Booking.id == booking_id).first()
    if not booking:
        raise HTTPException(status_code=404, detail="Booking not found")
    trip = db.query(Trip).filter(Trip.id == booking.trip_id).first()
    driver = db.query(User).filter(User.id == trip.driver_id).first() if trip else None
    return {
        "id": booking.id,
        "trip": {
            "id": trip.id,
            "origin": trip.origin,
            "destination": trip.destination,
            "departure_time": trip.departure_time,
            "driver_id": trip.driver_id,
            "driver": {
                "id": driver.id,
                "full_name": driver.full_name,
            } if driver else None,
        },
        "seats_booked": booking.seats_booked,
        "total_price": booking.total_price,
        "status": booking.status,
        "created_at": booking.created_at,
    }


@router.put("/{booking_id}/confirm")
def confirm_booking(booking_id: int, db: Session = Depends(get_db)):
    booking = db.query(Booking).filter(Booking.id == booking_id).first()
    if not booking:
        raise HTTPException(status_code=404, detail="Booking not found")

    if booking.status != "paid":
        raise HTTPException(
            status_code=400, detail="Booking must be paid before confirmation"
        )

    booking.status = "confirmed"
    db.commit()
    return {"message": "Booking confirmed", "status": "confirmed"}


@router.put("/{booking_id}/cancel")
def cancel_booking(booking_id: int, db: Session = Depends(get_db)):
    booking = db.query(Booking).filter(Booking.id == booking_id).first()
    if not booking:
        raise HTTPException(status_code=404, detail="Booking not found")

    if booking.status == "cancelled":
        return {"message": "Already cancelled"}

    # Refund seats and money
    trip = db.query(Trip).filter(Trip.id == booking.trip_id).first()
    passenger = db.query(User).filter(User.id == booking.passenger_id).first()

    trip.available_seats += booking.seats_booked
    passenger.balance += booking.total_price

    booking.status = "cancelled"
    db.commit()
    return {"message": "Booking cancelled and refunded"}
